### DISCUSSION TITLE
Future direction of VExim

### DISCUSSION THREAD

**Discussion by rimas-kudelis** [vexim/vexim2#275] [Ideas]
While updating the README after our discussions on #272 and other issues, I looked at [what alternatives Vexim has](https://alternativeto.net/software/postfix-admin/). Here are the ones that specifically caught my eye:
- [Sophimail](https://alternativeto.net/software/sophimail-webadmin/about/): has some sexy screenshots, seems quite capable, although their website is no longer reachable and the last commits in their [GitHub repo](https://github.com/sophimail/webadmin) are from 2019. Licensed under GPL 3.0.
- [Modoboa](https://alternativeto.net/software/modoboa/about/): packed with features (Sieve, Amavis, DMARC, Autoreply), liberally (ISC) licensed, Python-based (ping @runout-at), [maintained](https://github.com/modoboa/modoboa/commits/master) (has a bunch of commits from 2021 and 2022), and even commercially supported. I was somewhat surprised that same app is used as a webmail client, caldav client and admin interface, but from looking at [the docs](https://modoboa.readthedocs.io/en/latest/), it seems to use plugins for much of the functionality, so you can probably just disable/not install these end-user facing features if you don't want them. 
- [mailcow: dockerized](https://alternativeto.net/software/mailcow/about/): sexy screenshots, [actively maintained](https://github.com/mailcow/mailcow-dockerized/commits/master), [lots of features](https://mailcow.github.io/mailcow-dockerized-docs/) (including U2F support for @Udera), even Dockerized out-of-the-box! The admin panel seems to use plain PHP without a framework (weird choice I'd say), but at least Twig is used for templates and the code is certainly much cleaner than ours. Licensed under GPL 3.0. Has commercial support and managed hosting options. 

We could also add to that @runout-at's [veximpy](https://gitlab.com/runout/veximpy), which is a near-complete reimplementation of VExim in Python.

With all these options out there, and the general lack of interest in this project (even from its maintainers, including myself), and issues like #272, I'm actually tempted to just archive this repository and point people towards these more active alternatives.

What do you guys think? Tagging @avleen, @Udera, @runout-at, @gldickens3 and @corux here, but anyone should feel free to weigh in.

**Comment by runout-at** [comment]
I had a look at modoboa. It has a lot of features we (still) don't have in Vexim. But Vexim has in it's features much more functionality and it seems that modoba is trying to provide a little bit of everything. Personally I'd go for separate tools for some of the features.

[sogo](https://www.sogo.nu/about.html#/features) could be interesting - I don't know if it has the admin tools we need. Maybe good to use together with Vexim.

As mailcow is docker based it's not an option for me.

About closing this repo...
Even if somebody goes for another frontend, Vexim has a lot of config to connect Exim to a database and this could be useful for any frontend. Most applications don't bring anything for Exim.

**Reply by rimas-kudelis** [reply]
Oh, I thought you copied these config files to your repo.

If you haven't, that's cool. We could proceed with removing the UI, or splitting it into its own repository, and keep this one just for configs then.

I was also contemplating what would be a good way to maintain correct version association between the schema repo and the UI repo. One way would be to actually version the database with each small change, but I'm thinking perhaps it would be easier for UI to just reference the correct git commit as a submodule? This way each git commit would be a new "version" in a way. WDYT?

**Reply by rimas-kudelis** [reply]
Also, sogo seems to be something entirely different from what we are doing. Mailcow even ships a container with sogo.

**Reply by runout-at** [reply]
Yes. My intention was to split the GUI from the Exim configs and the database

I already had similar thoughts about syncing GUI/DB/Exim-CFG. Not sure what the best way would be.
Maybe using Releases...

**Comment by Udera** [comment]
What I like about the vexim approach was that it is minimalistic and flexible. Just perfect, if you need a tool to configure virtual mail accounts. And you can extend it like you wish to have whatever webmail you prefer, etc.

However nowadays, setups are a bit complicated, you can manually put some features (DKIM, etc.). but some of them could be  integrated without problems. The setup procedure is not very streamlined. However, calendar and stuff, that's not really part of a mail server...

Advantage is in theory, you have little code and you only have to make sure to sanitize the user input correctly (but you don't depend on huge frame works or 3rd-party code that needs to be maintained). I am not sure if it is possible to keep an easy low-maintenance solution...
I remember some time ago, we did also some ping pong with Exim4U, which uses vexim at the base and implements a few more features.

Lately, I tried to integrate Mailman3 which is already using quite a lot of python3. So with veximpy, we could do everything in python. I haven't used it for web applications yet.

**Reply by runout-at** [reply]
Exim4U tries to do put all the components for a email system together. As I like to split services to different VMs, it was not a solution for me.

I had a look at Mailman3 some years ago but back then were still some problems in the official Debian package (I wrote some reports) and it needs lots of RAM as it uses the Django framework (compared to Mailman2) - I should give it a try again and I'm already thinking about integration in veximpy.

**Comment by MrSleeps** [comment]
I realise this is an old discussion but I feel VExim2 shouldn't be archived. It needs a fair bit of work (no offence) to bring it up to modern standards and a fair bit of work needs to be done on the install/guides.

I'm more than happy to start helping but looking at the pull requests, some are outstanding from a few years ago, makes me wonder if it's worth it if PRs are just going to sit there? (again, no offence).

**Comment by avleen** [comment]
I'm super happy for anyone who would like to contribute!
I haven't had time to do this for some time now. And I stopped running exim
locally a while ago too.

But I'm happy to start pumping out a bunch of improvements that could be
helpful here.
I think the first and most important thing we need it to improve the state
of testing.
We need a proper test setup which can be used to validate all changes.
It's a really good use case for Claude or Codex if anyone would like to
take a shot.

**Comment by MrSleeps** [comment]
More than happy to help out. I would say the most important thing is working out what direction it's heading? There's been a lot of posts asking about or suggesting things, opinions voiced and even some PRs but no overall decision made.

For me, the important things would:
Documentation, it's a great starter but lacking in places
Modernising the SQL @rimas-kudelis updated it in a PR a while back if I remember?
Modernising the external programs that VExim uses, Dovecot does LMTP and has done for a while, RSpamd, MM3..
Securing Exim, those pesky pipes again but also Exim has changed a lot since this was first released.
Web UI, the old one is what, 20 years old? Veximpy is amazing but not everyone knows python..
Direction, where it's going, who wants to work on it

Testing, I have a server which I/we can use for testing. It's what I was using when I was rebuilding the UI.  Obviously, I can't give out root access (being a work test server) but I can easily swap out configs from ours to vexims.

I realise I've not contributed much over the years (have done a few lately since I noticed it wasn't dead) so my opinion is probably not worth much.

But there's no reason why this great project can't be even better.

**Comment by runout-at** [comment]
Some very old PRs from me are still open and I don't know how to change some issues in the PRs because the branches are not available anymore. I'm not a git-guru - maybe it's possible (please teach me how).

When I was tied in as a project member because of veximpy I didn't merge my own PRs because I'd like somebody else to do a review and have an agreement about the changes with the others.

There are not many alternatives to Vexim. Exim4U is trying to do everything very monolithic and I don't like that approach.

What I miss in Vexim is the possibility to use a LDAP backend and multiple backends.

For using veximpy not much python knowledge should be necessary.

MM3 I tried several times when it was released but it was really painful. If you just need a slim list server without all the fancy stuff I'd recommend [MLMMJ](https://github.com/vexim/vexim2/wiki/Server-configuration:-MLMMJ-(Mailing-Lists)).

**Comment by rimas-kudelis** [comment]
For now, the smaller and less complex the PR, the better its chances are of being accepted quickly.

I was (and still am, to a degree) reluctant to merge any database schema changes, until we devise a way to version it properly. I tied this versioning to a rewrite of the PHP interface, but I guess this wouldn't be nice towards alternative frontends like VeximPy, so perhaps a better idea would be to just introduce a new table to store various Vexim variables like that and make something like `vexim_db_schema_version` the first variable in that table.

As for other complex things...

1. I'm still unsure how we should approach distributing the daemon configuration files/changes to them, especially when they might have to differ between daemon versions. On one hand, shipping a working config is nice, but on the other, we end up burdened with keeping that file up-to-date, applying latest best practices, etc.
2. At the same time, we aren't even providing any sort of a "turnkey solution" which the admin could "just run" in their preferred virtualization/containerization environment with minimal effort. A 10-page installation manual, which assumes that everything will be set up on the same physical machine, was normal 20 or 30 years ago, but is it really still normal today? I personally would like to see VExim and Exim4U converging into one project, where we would provide both the building blocks and a complete solution "reference implementation", if you like.
3. Regarding LDAP backend: I still fail to grasp how we would tie into it. What exactly would you use VExim for if you managed your domains and users externally via LDAP? As I see it, VExim is a database schema, some daemon configs using that schema, and a web UI for managing entries in that schema. If you take that schema away, does anything useful/usable remain?
4. As for mailing lists, our integration with Mailman is currently next to none: we include some mailman-related lines in the config files and have a page which allows quicker access to mailman list management if `$mailmanroot` is defined. I find it hard to label that as "integration". IMO, an integration like this is of very little value and easy to replace with something more generic (e.g. configurable custom links for the WebUI). The reason why we even have it is probably because we're shipping a full config file instead of just snippets that we actually need (see point 1). Unless we can extend/supplement whatever Mailman or MLMMJ are doing in a useful way, I'd say we don't need any of these integrations.

**Comment by MrSleeps** [comment]
DB versioning, you can do that with a simple table (as you said) in the database.

```
CREATE TABLE schema_migrations (
  version TEXT PRIMARY KEY,
  applied_at TIMESTAMP DEFAULT NOW()
);
```

Have a layout in the repo something like this:

```
/db/migrations
  0001_init.sql
  0002_add_dkim_table.sql
  0003_modify_domains_table.sql
```
  
Then when you need to alter the schema, you create a new branch (v3.1 for example), add the sql migration to the db folder so it becomes something like

```
/db/migrations
  0001_init.sql
  0002_add_dkim_table.sql
  0003_modify_domains_table.sql
  0004_a_new_migration.sql
```
  
Then you need a script that gets the list of migrations in the folder (in order) and then compares those to the migrations table in the vexim database. If it's not listed in the db it runs the migration. Takes a bit of organisation but that is basically the way Laravel handles it's migrations.

Migrations can be safeguarded by checking if it exists (https://mariadb.com/docs/server/reference/sql-statements/data-definition/alter/alter-table) but that only works in MariaDB, Postgres has "ADD COLUMN IF NOT EXISTS" if I remember rightly. MySQL.. Not so lucky, that's manual logic involved (unless something has changed since I last used MySQL).

For configs (point 1), the simplest way (IMO) is to make a folder with the release number (v3.0, v3.1 etc), copy over all the files in that. Edit them (add the updates) or add the new files and that gets merged. So you end up with a million folders under config, but each folder has the full configs for that version. The release notes should include a list of the changes. Install guides etc get put in that folder too.

So for example off the root

```
/configs
   /v3.0
     /exim
       /conf.d
	     /main
		  /00_existing_file
   /v3.1
     /exim
       /conf.d
	     /main
		  001_existing_file
		  002_new_file
```
		  
You end up with a lot of files but it's all versioned. Main docs refer to the latest versions. Old versions never get the PR accepted if the old configs/docs have been modified. I'm sure there's a better way of doing it but that's a very simple, can see exactly what has changed.

Or, you head down the evil route of patch files..

If you use SemVer it's pretty easy for the user to work out what is going on? You have vMAJOR.MINOR.PATCH:
MAJOR – Breaking changes (incompatible config changes).
MINOR – New features (backward compatible).
PATCH – Bug fixes (backward compatible).

2) There's no reason why we can't. If it's stable, tested and happy with it then github actions can build docker images (but then, do you really want to head down the route of maintaing docker images?). What we don't want to end up doing is what others have done, become the all in 1 and then get hammered with security issues (naming no names)

3) With you on that one, not sure how it ties in with VExim.

4) You can do this via plugins for the web interface. So you have the main web ui, then you have a plugin repo for say MM3 (I only mentioned that because the original had mailman mentioned). If the user wants MM3 support, they (assuming a php web ui) run composer require vexim/webui-mm3 and that installs that plugin and gets configured via the web ui, the user then has mailmain 3 support in the web-ui and the daemon config for that is provided in the /configs/version directory.

**Reply by rimas-kudelis** [reply]
Our migrations are currently similar to what you described, except we don't track whether they've been applied.

As for configs, I'm not sure why we'd want to keep all of their versions next to each other assuming they're already versioned by git.

Re Docker: I'm not really sure I follow you. My vision is that the most important image we'd probably be building would be the one with Exim using our configuration scheme. Other services (database, web server, web app code) would use their own containers when possible. It would all be tied together by Docker Compose or Kubernetes, or whatever (I myself only have experience with Docker Compose though). Our images would be based on stock ones, only adding what's needed for our own project to work and only when necessary (like, if an image can be configured entirely via the environment, we could just use the stock one). If you look at the current `docker-compose.yaml`, you should get the gist of what I mean, except there are no mail server containers configured.

I was thinking about plugins too. One risk I see in that approach is the possibility of different webUIs ending up with different supported features and plugin sets. But I guess it's neither something we could avoid, nor something to fret about too much, assuming we don't become **THE** virtual mail solution overnight. I guess we'll figure these things out as we go.

Finally, a warning: I'm seriously overwhelmed with other things at the moment. I won't have the willpower or motivation to review all patches, especially the bigger ones, especially ones I'd have to test by tinkering with my Exim installation. Help on that front would be appreciated.

**Reply by MrSleeps** [reply]
The migrations table is what is missing, that's a pretty simple thing to implement. When I'm back at work I'll have a snoop through what we use and see what I can pull together from that.. Though @runout-at is more of a python person than me so is way more suited to that. As for the docs, was just an idea and it gives the user instant full configuration files for the current version (we use tags at work for which I am constantly forgetting to switch to to track changes - v1 tag, v2 etc)

Docker, yeah I didn't explain that too well. There's no reason why we can't have a docker with a working version of VExim (Exim, SQL, Dovecot etc). But keeping it up to date is the issue. Just look what happened to the popular all in 1 mail docker image, they've been hit with a fair few (CVE-2025-25198 being one of them). Docker is the easier option out of kubernetes etc, more people know it and run it.. It's a fair amount of work to keep them up to date, bug free and safe. 

Plugins are a great way to go, with my web ui it prepends the table name with vw_ to differentiate between my stuff and the core VExim stuff. We could also do the same prefixing with "plugins" like mm3, rspamd, mmlmj.

I have the work test server I can use or my own personal mail server which I'm not too fussed if it breaks for an hour or two (work one is a bit more of an issue if I break it). So am more than happy to help out checking PRs but it would need someone to eventually merge them into the branch.

**Reply by rimas-kudelis** [reply]
Having a migrations table is pretty simple, it's just I'd like to figure out how to make it client-agnostic, how to support downstream client-specific changes etc. In this context, a "client" is a WebUI or TUI or GUI or anything else.

On the other hand, Core migrations should remain plain SQL files if we want to be client-agnostic, so maybe I'm overthinking things here and we should just continue doing what we're doing already.

I'm also a bit unsure about how we should link multiple repos together, but perhaps it's not really a problem Core should be solving either. Perhaps just adhering to semantic versioning should be enough.

**Reply by MrSleeps** [reply]
VExim is the base for everything, and anything else has to work with what VExim has chosen and whatever changes we make (I'm talking as if I maintain it, you get what I mean). It's not really on VExim to bend to whatever downstream clients (UIs, TUIs, etc) decide to do unless it's something that gets proposed back properly via PRs and lands in core.

For client specific stuff, I think it's on them to handle their own migrations. Like in my case my UI prefixes vw_ onto everything ui related and keeps it as separate as I can reasonably make it. That said, I've also already pulled in DB changes from unmerged PRs so I've kind of broken my own rule there.

On linking repos, I think semVer is probably enough. Clients pin a version, they get a known set of behavior, and everyone's on the same page. If they've diverged from core, that's kind of their problem to untangle rather than something core should try to accommodate.

And yeah, I'm fully aware I'm commenting on this while also having done some of the things I'm saying we shouldn't do, so take it with a pinch of salt but I still think this is the right direction.

Edit: I have the feeling I totally missed the point of what you were asking..

**Comment by MrSleeps** [comment]
On another note, a web ui.. Do you need an official one? Veximpy is out there, I've just released a php version (which is running on my personal mail server). Surely just provide links to those repos and let the user choose (or roll their own?). An API only version wouldn't be too difficult to roll out (like what power dns does).

**Reply by rimas-kudelis** [reply]
Yeah, extracting the WebUI into its own repo is also an idea I've been entertaining for quite a while. Now with two already existing new reimplementations, "extracting" might just mean `rm -Rf`.

It would be nice if the different UIs kept supporting same database table structure for VExim "stock" tables though and coordinated changes to remain interoperable.

**Reply by MrSleeps** [reply]
Yeah, mine has taken some of the PRs and integrated them already (DKIM and RSpamd support for example). More than happy to change mine to fit with the "standard" VExim. Currently part way through writing a plugin for mine that supports PowerDNS dns updates (so once finished you can get an idea of what is possible with plugins).

I'll be updating mine as and when it needs it because, as I said, I'm now using it on my personal server.

**Comment by runout-at** [comment]
Vexim could become the base project to have a definition for the database and provide the Exim config on which different fontends could sit and be interchangeable. Even other software like MM3, MLMMJ,... could be served with some common API specifications eg. for subscribing/managing  virtual users, ... by those frontends (just a thought what could be).

About a CLI version: I was thinking about that when writing veximpy. Should not be too hard to extend the functionality but I didn't need it until now.

**Reply by MrSleeps** [reply]
I agree, VExim should be the backend stuff. Users have a choice of front ends now.

Configs, yes. Though my front end will never support anything that runs with pipes unless the backend is tightened up. Releasing any software/ui with a known and valid security hole is just crazy. It also needs to be extended/updated, LMTP is a must and allowing extra replacement software (mlmmj instead of mm3, rspam instead of spamassassinm etc)

API spec is also a really good idea, but we need to be flexible and open up (for example, I will always support RSpamd because I use it, if the API drops support for anything related to that then my ui will always drift from it).

**Reply by rimas-kudelis** [reply]
IMO, an API is essentially just another type of front-end. The core probably doesn't need to implement one, since it too can be written in any language and, in turn, be too unfamiliar or too unsafe for some.

**Reply by MrSleeps** [reply]
It's just an option to let people roll their own front end, if they can see what the "official" does, it's a base for them to start with. I'm in the middle of building one to allow remote control of VExim (1 single ui, control of multiple (remote) servers to avoid database replication hassles)..

**Reply by MrSleeps** [reply]
> could be served with some common API specifications eg. for subscribing/managing virtual users, ... by those frontends (just a thought what could be).

I've started work on mine now, very rough draft for domains is [here](https://mrsleeps.github.io/vexim-web-docs/api/domains.html)

**Comment by avleen** [comment]
I've started building a proper test harness. This will bring together unit
tests and a local exim config to do integration tests. We'll start by
making sure we have a set of tests which pass, so that we can make changes
more confidently.

From there we can start making the changes we need!

**Reply by MrSleeps** [reply]
I'll have a chat with people at work, see if I can get them to donate a vm for testing, just need to find the least risk averse person to talk to. Seems fair seeing as our entire mail setup was loosely based on VExim.
